src/QsCompiler/Transformations/CallGraphWalker.cs

```diff
@@ -17,6 +16,259 @@ namespace Microsoft.Quantum.QsCompiler.Transformations.CallGraphWalker
+    /// in which all nodes can reach all other nodes in the subgraph.
+    ///
+    /// The algorithm begins by using Tarjan's algorithm to find the SCCs of the graph.
+    /// Each SCC is processed for cycles, using the smallest-id node of the SCC as the
+    /// starting node. Then the starting node is removed from the graph, and Tarjan is used
+    /// again to find SCCs of the new graph, and the next SCC is processed. This repeats
+    /// until there are no more SCCs (and no more nodes) left in the graph.
+    ///
+    /// The algorithm has time complexity of O((n + e)(c + 1)) with n nodes, e edges,
+    /// and c elementary cycles.
+    ///
+    /// The starting point for this algorithm in this class is the GetAllCycles method.
+    /// </summary>
+    internal class JohnsonCycleFind
+    {
+        private Stack<(HashSet<int> SCC, int MinNode)> SccStack = new Stack<(HashSet<int> SCC, int MinNode)>();
+
+        /// <summary>
+        /// Johnson's algorithm for finding all cycles in a graph.
+        /// This returns a list of cycles, each represented as a list of nodes. Nodes
+        /// for this algorithm are integers. The cycles are guaranteed to not contain
+        /// any duplicates and the last node in the list is assumed to be connected
+        /// to the first.
+        /// </summary>
+        public List<List<int>> GetAllCycles(Dictionary<int, List<int>> graph)
+        {
+            // Possible Optimization: Each SCC could be processed in parallel.
+
+            // Possible Optimization: Getting the subgraph for an SCC only needs
+            // to consider the parent graph, not the full graph. A recursive strategy
+            // may be better than using a stack object.
+
+            var cycles = new List<List<int>>();
+
+            PushSCCs(graph);
+            while (SccStack.Any())
+            {
+                var (scc, startNode) = SccStack.Pop();
+                var subGraph = GetSubGraph(graph, scc);
+                cycles.AddRange(GetSccCycles(subGraph, startNode));
+
+                subGraph.Remove(startNode);
+                foreach (var (_, children) in subGraph)
+                {
+                    children.Remove(startNode);
+                }
+
+                PushSCCs(subGraph);
+            }
+
+            return cycles;
+        }
+
+        /// <summary>
+        /// Uses Tarjan's algorithm for finding strongly-connected components, or SCCs of a
+        /// graph to get all SCC, order them by their node with the smallest id, and push them to
+        /// the SCC stack.
+        /// </summary>
+        private void PushSCCs(Dictionary<int, List<int>> graph)
+        {
+            var sccs = TarjanSCC(graph).OrderByDescending(x => x.MinNode);
+            foreach (var scc in sccs)
+            {
+                SccStack.Push(scc);
+            }
+        }
+
+        private Dictionary<int, List<int>> GetSubGraph(Dictionary<int, List<int>> inputGraph, HashSet<int> subGraphNodes) =>
```

Comment:
>GetSubGraph [](start = 43, length = 11)

A comment here would really help me understand what is happening. What kind of subgraph are we trying to get? It would also help to have a more descriptive name for this function. I don't care how long it will become - it is only used once in the code, but it would help understanding.

Reply:
I added a doc comment, renamed the method, and added some comments to the code, but I'm not sure it is any clearer.

Reply:
The comment definitely helps. Thanks. A better name (as I understand now) would be GetSubGraphLimitedToNodes.

---
In reply to: [405006096](https://github.com/microsoft/qsharp-compiler/pull/396#discussion_r405006096) [](ancestors = 405006096)